### PR TITLE
Stabilize CoroutineStart.ATOMIC

### LINKED ISSUE #4151: Consider stabilizing `CoroutineStart.ATOMIC`
The `CoroutineStart.ATOMIC` option was marked experimental almost 6 years ago as part of version 0.30.0. Is there a strong reason that the option is still experimental, or could it be considered stable by now?

### PR DESCRIPTION
**Pull request by zuevmaxim**
Fix #4151

### CODE REVIEW

File: kotlinx-coroutines-core/common/src/CoroutineStart.kt [hunk 1]
```diff
@@ -48,10 +48,29 @@ public enum class CoroutineStart {
      * Atomically (i.e., in a non-cancellable way) schedules the coroutine for execution according to its context.
      * This is similar to [DEFAULT], but the coroutine cannot be cancelled before it starts executing.
      *
+     * The coroutine stated with [ATOMIC] is guaranteed to start execution even if its [Job] was cancelled.
```
Comment (qwwdfsad) [inline kotlinx-coroutines-core/common/src/CoroutineStart.kt:51]: Probably you've meant `started` instead of `stated`?

**Review by qwwdfsad — APPROVED** [review]

File: kotlinx-coroutines-core/common/src/CoroutineStart.kt [hunk 2]
```diff
@@ -48,10 +48,29 @@ public enum class CoroutineStart {
...
+     *     launch(start = CoroutineStart.ATOMIC) {
+     *         try {
+     *             // produce elements
+     *         } finally {
+     *             channel.close()
+     *         }
+     *     }
+     * ```
+     *
+     * This is a **delicate** API. The coroutine starts execution even if its [Job] is canceled before starting.
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/CoroutineStart.kt:65]:
```suggestion
     * This is a **delicate** API. The coroutine starts execution even if its [Job] is cancelled before starting.
```
Both versions are correct, but we use "cancelled" throughout the codebase.

File: kotlinx-coroutines-core/common/src/CoroutineStart.kt [hunk 3]
```diff
@@ -48,10 +48,29 @@ public enum class CoroutineStart {
...
+     * However, the resources used in within a coroutine may rely on the cancellation mechanism,
+     * and cannot be used after the [Job] cancellation. For instance, in Android development, updating a UI element
+     * is not allowed if the coroutine's scope, which is tied to the element's lifecycle, has been canceled.
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/CoroutineStart.kt:68]:
```suggestion
     * is not allowed if the coroutine's scope, which is tied to the element's lifecycle, has been cancelled.
```
Ditto.

File: kotlinx-coroutines-core/common/src/CoroutineStart.kt [hunk 4]
```diff
@@ -48,10 +48,29 @@ public enum class CoroutineStart {
      * Atomically (i.e., in a non-cancellable way) schedules the coroutine for execution according to its context.
      * This is similar to [DEFAULT], but the coroutine cannot be cancelled before it starts executing.
      *
+     * The coroutine stated with [ATOMIC] is guaranteed to start execution even if its [Job] was cancelled.
+     * This [CoroutineStart] option can be used to ensure resources' disposal in case of cancellation.
+     * For example, closing a channel used by a producer:
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/CoroutineStart.kt:53]: Is this a complete sentence? Maybe explain the idea more verbosely, like "For example, this `producer` guarantees that the channel will be eventually closed, even if the coroutine scope is cancelled before `producer` is called:"?

File: kotlinx-coroutines-core/common/src/CoroutineStart.kt [hunk 5]
```diff
@@ -48,10 +48,29 @@ public enum class CoroutineStart {
...
+     *         try {
+     *             // produce elements
+     *         } finally {
+     *             channel.close()
+     *         }
+     *     }
+     * ```
+     *
+     * This is a **delicate** API. The coroutine starts execution even if its [Job] is canceled before starting.
+     * However, the resources used in within a coroutine may rely on the cancellation mechanism,
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/src/CoroutineStart.kt:66]:
```suggestion
     * However, the resources used within a coroutine may rely on the cancellation mechanism,
```
A typo?

**Review by dkhalanskyjb — CHANGES_REQUESTED** [review]

**Review by dkhalanskyjb — APPROVED** [review]

**Pull request merged**